controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStoreHelper.java

```diff
@@ -0,0 +1,403 @@
+
+@Slf4j
+public class PravegaTablesStoreHelper {
+    private static final int NUM_OF_RETRIES = 15; // approximately 1 minute worth of retries
+    private final SegmentHelper segmentHelper;
+    private final ScheduledExecutorService executor;
+    private final Cache cache;
+    private final AtomicReference<String> authToken;
+    private final AuthHelper authHelper;
+    
+    @lombok.Data
+    @EqualsAndHashCode(exclude = {"fromBytesFunc"})
+    private class TableCacheKey<T> implements Cache.CacheKey {
+        private final String scope;
+        private final String table;
+        private final String key;
+
+        private final Function<byte[], T> fromBytesFunc;
+    }
+    
+    public PravegaTablesStoreHelper(SegmentHelper segmentHelper, AuthHelper authHelper, ScheduledExecutorService executor) {
+        this.segmentHelper = segmentHelper;
+        this.executor = executor;
+
+        cache = new Cache(x -> {
+            TableCacheKey<?> entryKey = (TableCacheKey<?>) x;
+
+            // Since there are be multiple tables, we will cache `table+key` in our cache
+            return getEntry(entryKey.getScope(), entryKey.getTable(), entryKey.getKey(), entryKey.fromBytesFunc)
+                                   .thenApply(v -> new VersionedMetadata<>(v.getObject(), v.getVersion()));
+        });
+        this.authHelper = authHelper;
+        this.authToken = new AtomicReference<>(authHelper.retrieveMasterToken());
+    }
+
+    public <T> CompletableFuture<VersionedMetadata<T>> getCachedData(String scope, String table, String key, Function<byte[], T> fromBytes) {
+        return cache.getCachedData(new TableCacheKey<>(scope, table, key, fromBytes))
+                    .thenApply(this::getVersionedMetadata);
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T> VersionedMetadata<T> getVersionedMetadata(VersionedMetadata v) {
+        // Since cache is untyped and holds all types of deserialized objects, we typecast it to the requested object type
+        // based on the type in caller's supplied Deserialization function. 
+        return new VersionedMetadata<>((T) v.getObject(), v.getVersion());
+    }
+
+    public void invalidateCache(String scope, String table, String key) {
+        cache.invalidateCache(new TableCacheKey<>(scope, table, key, x -> null));
+    }
+
+    public CompletableFuture<Void> createTable(String scope, String tableName) {
+        log.debug("create table called for table: {}/{}", scope, tableName);
+
+        return Futures.toVoid(withRetries(() -> segmentHelper.createTableSegment(scope, tableName, authToken.get(), RequestTag.NON_EXISTENT_ID),
+                () -> String.format("create table: %s/%s", scope, tableName)))
+                .whenCompleteAsync((r, e) -> {
+                    if (e != null) {
+                        log.warn("create table {}/{} threw exception", scope, tableName, e);
```

Comment:
What's the consequence of the create table failing? We are logging a warn message, but what happens to the caller?

Reply:
It depends on the caller and its exception handling for cause for failure. If its a connection failure, its a retryable failure and caller can retry or reschedule their work. 

For example, for createScope, if a `createTable` call fails after exhausting retries in the store, the requested grpc call for createScope is failed. 
If the `createTable` is called from one of the workflows like say `scale`, then create table is effectively retried indefinitely